### PR TITLE
[backup] TransactionRestoreController: replay transactions

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation
so that with a state snapshot and transaction backup on top of it we can
recover any state that's good for state-sync to boot.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

yes

## Test Plan

unit tests
## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/libra/website, and link to your PR here.)

### CODE REVIEW

File: storage/backup/backup-cli/src/bin/db-backup.rs [hunk 1]
```diff
@@ -68,19 +91,46 @@ async fn main() -> Result<()> {
...
+                    BackupType::EpochEnding { opt, storage } => {
+                        EpochEndingBackupController::new(
+                            opt,
+                            global_opt,
+                            client,
+                            storage.init_storage().await?,
+                        )
+                        .run()
+                        .await
+                        .map(|hdl| println!("Epoch ending backup success. Manifest: {}", hdl))
```
Comment (wqfish) [inline storage/backup/backup-cli/src/bin/db-backup.rs:106]: `hdl`?

Comment (msmouse) [reply storage/backup/backup-cli/src/bin/db-backup.rs:106]: manifest_HanDLe I meant.. let's do "m" then

File: storage/libradb/src/backup/restore_handler.rs [hunk 2]
```diff
@@ -122,4 +140,23 @@ impl RestoreHandler {
...
     }
+
+    pub fn get_tree_state(&self, num_transactions: LeafCount) -> Result<TreeState> {
+        let frozen_subtrees = self
+            .ledger_store
+            .get_frozen_subtree_hashes(num_transactions)?;
+        let state_root_hash = if num_transactions == 0 {
+            self.state_store
+                .get_root_hash_option(PRE_GENESIS_VERSION)?
+                .unwrap_or(*SPARSE_MERKLE_PLACEHOLDER_HASH)
```
Comment (lightmark) [inline storage/libradb/src/backup/restore_handler.rs:151]: why not expect?

File: storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs [hunk 3]
```diff
@@ -82,9 +87,31 @@ impl EpochEndingRestoreController {
...
+                    lis[_end].ledger_info().epoch(),
+                    lis[_end].ledger_info().version(),
+                    self.target_version,
+                );
+                end = _end;
+            }
+
             // write to db
-            self.restore_handler.save_ledger_infos(&lis)?;
+            if end != 0 {
```
[the remark this answers is not in the document]
Comment (msmouse) [reply storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs:99]: "begin == end" means an empty range, again it's natural to c++/STL guys.

File: storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs [hunk 4]
```diff
@@ -27,13 +27,18 @@ pub struct StateSnapshotRestoreOpt {
 pub struct StateSnapshotRestoreController {
     storage: Arc<dyn BackupStorage>,
     restore_handler: Arc<RestoreHandler>,
+    // State snapshot restore to this version
     version: Version,
     manifest_handle: FileHandle,
+    // Global "target_version" for the entire restore process, if `version` is newer than this,
+    // nothing will be done.
```
[the remark this answers is not in the document]
Comment (msmouse) [reply storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs:31]: the snapshot will be restored

File: storage/backup/backup-cli/src/backup_types/transaction/restore.rs [hunk 5]
```diff
@@ -72,14 +76,35 @@ impl TransactionRestoreController {
...
+                )?;
+                first_chunk = false;
+            }
+
+            if last < chunk.last_version {
+                break;
+            }
+        }
+
+        if self.target_version < manifest.first_version {
```
[the remark this answers is not in the document]
Comment (msmouse) [reply storage/backup/backup-cli/src/backup_types/transaction/restore.rs:195]: I didn't like these printing logic scattered.
But whatever.. let's move these two up ahead and change "ignored" to "will be ignored".

File: execution/executor/src/lib.rs [hunk 6]
```diff
@@ -152,38 +152,81 @@ where
...
             first_txn_version <= num_committed_txns,
             "Transaction list too new. Expected version: {}. First transaction version: {}.",
             num_committed_txns,
             first_txn_version
         );
-        Ok((
-            num_committed_txns - first_txn_version,
-            num_committed_txns as Version,
-        ))
+        if first_txn_version + txn_list_with_proof.transactions.len() as u64 <= num_committed_txns {
```
[the remark this answers is not in the document]
Comment (msmouse) [reply execution/executor/src/lib.rs:187]: it's probably not intuitive either way, (especially to different eyes :P) I'll do
```
let version_after_txn_list = first_txn_version + txn_list_with_proof.transactions.len();
if version_after_txn_list >= num_commited_txns {
```

File: execution/executor/src/lib.rs [hunk 7]
```diff
@@ -499,13 +500,15 @@ where
...
+            V::execute_block(transactions.clone(), &state_view)?
         };
 
         // Since other validators have committed these transactions, their status should all be
         // TransactionStatus::Keep.
         for output in &vm_outputs {
-            if let TransactionStatus::Discard(_) = output.status() {
+            if let TransactionStatus::Discard(e) = output.status() {
+                println!("discard: {}", e);
+                println!("discard: f:{}", first_version);
```
[the remark this answers is not in the document]
Comment (msmouse) [reply execution/executor/src/lib.rs:511]: debug message.. gonna remove

File: storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs [hunk 8]
```diff
@@ -82,9 +87,31 @@ impl EpochEndingRestoreController {
                 next_epoch += 1;
             }
 
+            let mut end = lis.len();
+            if let Some(_end) = lis
+                .iter()
+                .position(|li| li.ledger_info().version() > self.target_version)
```
[the remark this answers is not in the document]
Comment (msmouse) [reply storage/backup/backup-cli/src/backup_types/epoch_ending/restore.rs:87]: Back in the C++/STL days "[begin, end)" half-close ranges are so natural and i got used to them. "end" representing one pass the "last" and "end == len()" means not found. so yeah ">" is intended here.

File: storage/libradb/src/backup/restore_handler.rs (hunk 2 above)
Comment (msmouse) [reply storage/libradb/src/backup/restore_handler.rs:151]: ? when there's no PRE_GENESIS_VERSION root, we need to fallback to the default hash.

File: execution/executor/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (lightmark) [reply execution/executor/src/lib.rs:187]: when it comes to this problem, my concern is always bout overflow.

File: storage/libradb/src/backup/restore_handler.rs (hunk 2 above)
Comment (lightmark) [reply storage/libradb/src/backup/restore_handler.rs:151]: I think I am confused cuz I thought the `PRE_GENESIS_VERSION` will be always pre-created.

File: storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (lightmark) [reply storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs:31]: which one? I mean if `version < target_version`, it is not clear to me from the desc.

**Comment by lightmark** [comment]
Huge PR but LGTM. Please add necessary comments and then land.

File: storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (msmouse) [reply storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs:31]: there's the comment above `version` saying that's the version to restore to..

[the remark this answers is not in the document]
Comment (msmouse) [reply storage/backup/backup-cli/src/backup_types/state_snapshot/restore.rs:31]: gonna add clarification.

File: execution/executor/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (msmouse) [reply execution/executor/src/lib.rs]: okay.. good point..

**Pull request closed without merging**